File: app/src/main/java/com/zacharee1/systemuituner/misc/BillingUtil.java
package com.zacharee1.systemuituner.misc;

import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.android.billingclient.api.BillingClient;
import com.android.billingclient.api.BillingClientStateListener;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.Purchase;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.PurchaseEvent;
import com.zacharee1.systemuituner.R;

import java.math.BigDecimal;
import java.util.Currency;
import java.util.List;
import java.util.Locale;

import static com.zacharee1.systemuituner.misc.Utils.isPackageInstalled;


public class BillingUtil
{
    private final Activity mActivity;
    private BillingClient mBillingClient;
//    private final List<String> mSkuList;

    public BillingUtil(Activity activity) {
        mActivity = activity;

        mBillingClient = BillingClient.newBuilder(mActivity).setListener(new PurchasesUpdatedListener()
        {
            @Override
            public void onPurchasesUpdated(int responseCode, List<Purchase> purchases)
            {
                if (responseCode == BillingClient.BillingResponse.OK && purchases != null) {
                    for (Purchase purchase : purchases) {
                        Answers.getInstance().logPurchase(new PurchaseEvent()
                            .putItemId(purchase.getSku())
                            .putSuccess(true)
                            .putCustomAttribute("orderId", purchase.getOrderId())
                            .putCustomAttribute("origJson", purchase.getOriginalJson())
                            .putCustomAttribute("packageName", purchase.getPackageName())
                            .putCustomAttribute("token", purchase.getPurchaseToken())
                            .putCustomAttribute("time", purchase.getPurchaseTime())
                            .putCustomAttribute("signature", purchase.getSignature()));
                        mBillingClient.consumeAsync(purchase.getPurchaseToken(), null);
                    }
                }
            }
        }).build();

        mBillingClient.startConnection(new BillingClientStateListener() {
            @Override
            public void onBillingSetupFinished(@BillingClient.BillingResponse int billingResponseCode) {
                Log.e("BillingResult", billingResponseCode + "");

                if (billingResponseCode == BillingClient.BillingResponse.OK) {
                    TextView ppTitle = mActivity.findViewById(R.id.paypal_title);
                    Button ppButton = mActivity.findViewById(R.id.paypal_button);

                    if (ppTitle != null) ppTitle.setVisibility(View.GONE);
                    if (ppButton != null) ppButton.setVisibility(View.GONE);
                } else if (billingResponseCode == BillingClient.BillingResponse.BILLING_UNAVAILABLE) {
                    LinearLayout gPlayD = mActivity.findViewById(R.id.google_play_donate);
                    TextView gPlayDT = mActivity.findViewById(R.id.google_play_donate_title);

                    if (gPlayD != null) {
                        gPlayD.setVisibility(View.GONE);
                    }

                    if (gPlayDT != null) {
                        gPlayDT.setVisibility(View.GONE);
                    }
                }
            }
            @Override
            public void onBillingServiceDisconnected() {
                // Try to restart the connection on the next request to the
                // In-app Billing service by calling the startConnection() method.
            }
        });
    }

    @SuppressWarnings("UnusedReturnValue")
    public int onDonateClicked(String skuId) {
        BillingFlowParams.Builder builder = BillingFlowParams.newBuilder()
                .setSku(skuId).setType(BillingClient.SkuType.INAPP);
        return mBillingClient.launchBillingFlow(mActivity, builder.build());
    }

    public static void onDonatePayPalClicked(Activity activity) {
        boolean labsInstalled = isPackageInstalled("com.xda.labs", activity.getPackageManager());
        Uri uri = Uri.parse(labsInstalled ? "https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=andywander@yahoo.com" : "https://forum.xda-developers.com/donatetome.php?u=7055541");
        Intent intent = new Intent(Intent.ACTION_VIEW, uri);
        activity.startActivity(intent);
    }
}

File: app/src/main/java/com/zacharee1/systemuituner/util/BillingUtil.kt
package com.zacharee1.systemuituner.util

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import com.android.billingclient.api.BillingClient
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingFlowParams
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.PurchaseEvent
import com.zacharee1.systemuituner.R


class BillingUtil(private val activity: Activity) {
    private val client: BillingClient

    init {
        client = BillingClient.newBuilder(activity).setListener { responseCode, purchases ->
            if (responseCode == BillingClient.BillingResponse.OK && purchases != null) {
                for (purchase in purchases) {
                    Answers.getInstance().logPurchase(PurchaseEvent()
                            .putItemId(purchase.sku)
                            .putSuccess(true)
                            .putCustomAttribute("orderId", purchase.orderId)
                            .putCustomAttribute("origJson", purchase.originalJson)
                            .putCustomAttribute("packageName", purchase.packageName)
                            .putCustomAttribute("token", purchase.purchaseToken)
                            .putCustomAttribute("time", purchase.purchaseTime)
                            .putCustomAttribute("signature", purchase.signature))
                    consumeAsync(purchase.purchaseToken)
                }
            }
        }.build()

        client.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(@BillingClient.BillingResponse billingResponseCode: Int) {
                if (billingResponseCode == BillingClient.BillingResponse.OK) {
                    val ppTitle = activity.findViewById<TextView>(R.id.paypal_title)
                    val ppButton = activity.findViewById<Button>(R.id.paypal_button)

                    if (ppTitle != null) ppTitle.visibility = View.GONE
                    if (ppButton != null) ppButton.visibility = View.GONE
                } else if (billingResponseCode == BillingClient.BillingResponse.BILLING_UNAVAILABLE) {
                    val gPlayD = activity.findViewById<LinearLayout>(R.id.google_play_donate)
                    val gPlayDT = activity.findViewById<TextView>(R.id.google_play_donate_title)

                    if (gPlayD != null) {
                        gPlayD.visibility = View.GONE
                    }

                    if (gPlayDT != null) {
                        gPlayDT.visibility = View.GONE
                    }
                }
            }

            override fun onBillingServiceDisconnected() {
                // Try to restart the connection on the next request to the
                // In-app Billing service by calling the startConnection() method.
            }
        })
    }

    private fun consumeAsync(token: String) {
        client.consumeAsync(token) { _, _ -> }
    }

    fun onDonateClicked(skuId: String): Int {
        val builder = BillingFlowParams.newBuilder()
                .setSku(skuId).setType(BillingClient.SkuType.INAPP)
        return client.launchBillingFlow(activity, builder.build())
    }

    companion object {
        fun onDonatePayPalClicked(activity: Activity) {
            val labsInstalled = activity.packageManager.isPackageInstalled("com.xda.labs")
            val uri = Uri.parse(if (labsInstalled) "https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=zachary.wander@gmail.com" else "https://forum.xda-developers.com/donatetome.php?u=7055541")
            val intent = Intent(Intent.ACTION_VIEW, uri)
            activity.startActivity(intent)
        }
    }
}
